src/tpm2.rs

```diff
@@ -0,0 +1,957 @@
+                self.sign_alg,
+                tf_path,
+                owner_pw,
+                aik_pw,
+                owner_pw
+            ),
+        };
+
+        let ret_out_map: Value = self
+            .run(cmd, None)
+            .and_then(|(ret_out, _)| Ok(serde_yaml::from_str(&ret_out)?))?;
+
+        match ret_out_map["load-key"]["namne"].as_str() {
+            None => {
+                return Err(KeylimeTpmError::new_tpm_rust_error(
+                    "Error: akname is missing in return output",
+                ))
+            }
+            Some(s) => self.update_tpmdata("aik_name", s)?,
+        }
+
+        match self.legacy_tools {
+            true => {
+                self.update_tpmdata(
+                    "aik_handle",
+                    std::str::from_utf8(&hex::decode("81010008")?)?,
+                )?;
+                self.get_pub_aik()?;
+            }
+            false => {
+                match ret_out_map["ak-persistent-handle"].as_str() {
+                    None => {
+                        return Err(KeylimeTpmError::new_tpm_rust_error(
+                            "ak-persistent-handle is missing",
+                        ))
+                    }
+                    Some(ak_handle) => {
+                        self.update_tpmdata("aik_name", ak_handle)?
+                    }
+                }
+
+                let aik_pw: String = self.read_file_output_path(tf_path)?;
+                self.update_tpmdata("aik_pw", &aik_pw)?;
+            }
+        }
+
+        self.update_tpmdata("aik_pw", &aik_pw)?;
+        Ok(())
+    }
+
+    // NVRAM io functions
+    fn read_ekcert_nvram(&self) -> Result<String, KeylimeTpmError> {
+        let tf = NamedTempFile::new()?;
+        let nvpath = self.temp_file_get_path(&tf)?;
+        let (ret_out, _): (String, HashMap<String, String>) =
+            self.run("tpm2_nvlist".to_string(), None)?;
+        let ret_out_value: Value = serde_yaml::from_str(&ret_out)?;
+        match ret_out_value["0x1c00002"]["size"].as_str() {
+            None => {
+                return Err(KeylimeTpmError::new_tpm_rust_error(
+                    "No EK certificate found in TPM NVRAM",
+                ))
+            }
+            Some(s) => {
+                let (_, f_out) = self.run(
+                    format!(
+                        "tpm2_nvread -x 0x1c00002 -s {} -f {}",
+                        s, nvpath
+                    ),
+                    Some(vec![nvpath]),
+                )?;
+                let ekcert = f_out.get(nvpath).ok_or_else(|| {
+                    KeylimeTpmError::new_tpm_rust_error(
+                        "failed to read nvram and retrieve ekcert.",
+                    )
+                })?;
+                Ok(base64::encode(ekcert))
+            }
+        }
+    }
+
+    fn read_key_nvram(&mut self) -> Result<String, KeylimeTpmError> {
+        // all error need to be handled by run function
```

Comment:
I don't understand this comment.

Reply:
Don't worry about this one, it was my note. Removed.